### ISSUE TITLE
Setting the Garden default grace time to a small value causes "unknown handle" error messages in resources

### ISSUE THREAD

**Issue by mbraku** [concourse/concourse#3054]
Hello,

I hope this is useful for you and we can fix the issues I'm seeing on our Concourse cluster together. Thank you for you help and sorry if that's not the right place to post this.

## Bug Report

I'm setting the default grace time for containers in Garden to a small value (something like 2m), so that containers are cleaned up more frequently (after some investigation we found out that we have a lot of check containers that are done but still lying around in Garden). But then we found out that this causes a lot of our resources to show an error saying "unknown handle: [UUID]".

## Steps to Reproduce

Set CONCOURSE_GARDEN_DEFAULT_GRACE_TIME to 2m (this is for the Kubernetes helm chart). This sets the default grace time for containers in Garden to two minutes.

## Expected Results

The resources should be checked normally even though we have cleaned up the previous check container.

## Actual Results

The resources error out producing an "unknown handle: [UUID]" error (see screenshot).

## Additional Context

We are running Concourse in a Kubernetes cluster using the official Kubernetes stable chart registry (https://github.com/helm/charts/tree/master/stable/concourse). We are running two worker pods with each pod running on their own VM exclusively (using node-selectors).

I have some logs from the worker that shows how the container is created and then deleted by guarding. I'd be happy to post it, but is there some output that I should better obfuscate before posting?

## Version Info

* Concourse version: 4.2.1
* Deployment type (BOSH/Docker/binary): Official Kubernetes stable chart
* Infrastructure/IaaS: Kubernetes
* Browser (if applicable): Chrome (but I don't think it's related)
* Did this used to work? I don't know

![screen shot 2019-01-16 at 10 21 21](https://user-images.githubusercontent.com/23443932/51220893-761fa280-197a-11e9-8239-85338f0722eb.png)

**Comment by mbraku** [comment]
I've also found out that running `watch fly -t [team] cr -r [pipeline]/[resource]` seems to eventually cause the check to work again. But I don't think this is a good way to fix this

**Comment by vito** [comment]
Thanks for the report! I've got a simple (kind of silly) answer: you shouldn't configure a default grace time. Concourse handles containers on its own via the ATC's database and has its own garbage collection logic for them - it's not up to Garden, and Concourse won't take nicely to containers being removed unexpectedly.

Later releases (5.0+) will make this flag harder to accidentally set - I can understand why it would look enticing if it seems like containers aren't going away quickly enough, but I suspect something else is going on here and that flag is definitely not the solution.

FYI, check containers intentionally stick around. You should expect to see one container for every resource in your pipeline. This is so that checks can cache state between checks, so that e.g. the `git` resource doesn't have to do a fresh clone every time. It could be fun thinking about a model where we don't re-use the containers and only re-use the state, though that's potentially a whole other topic for discussion. 🙂

**Comment by mbraku** [comment]
Thank you for your quick response!

Well, I definitely agree that configuring the default grace time isn't something that we want to be doing. It's just that if we don't define it, the amount of check containers on the workers is just too many to bear. We are running into various problems, from jobs not being started, to workers crashing, to other error messages (like https://github.com/concourse/concourse/issues/1620).

I understand that this is by design in order to optimize the system's performance. But when >90% of the worker's workload is dedicated to this and it cannot do the actual work that we need it to be doing, then it just seems to defeat the purpose of the exercise.

Could this be due to the fact that we are running the workers on Kubernetes, i.e. running Garden inside a Docker container? Maybe it is better to use a more official way such as BOSH (not that I have any experience with it)?

**Comment by vito** [comment]
@mbraku BOSH vs. K8s vs. Docker vs. binary shouldn't make any real difference here, so I wouldn't worry about that.

How many resources do you have across your pipeline, and how many workers do you have in order to support them? And how many physical machines? (Since running two workers on one machine doesn't help that much if it's already falling over due to load, for example.)

**Comment by vito** [comment]
@mbraku I'm gonna open an issue for this idea:

> It could be fun thinking about a model where we don't re-use the containers and only re-use the state, though that's potentially a whole other topic for discussion.

There's actually a lot of things this would simplify internally... I'll link it back to here once it's ready.

**Comment by mbraku** [comment]
I see, that's good to know.

We are currently running two worker pods on Kubernetes. And yeah, we've already noticed that the containers don't necessarily adhere to the CPU and memory limits of Kubernetes, so we made sure that both workers run on separate VMs exclusively. I cannot say anything about the physical machines, cause the Kubernetes cluster is managed by Azure.

We have >30 pipelines with >5 resources each. So it definitely makes sense that two workers are being overloaded by the amount of containers being created. If we put all pipelines on pause and just selectively unpause pipelines whenever we get a new git commit everything runs smooth as butter. But that's hardly an attractive solution to this issue.

Thank you for creating the new issue! We are very eager to use concourse at my company because it makes sense to use on a theoretical level, but the practical execution is currently a bit hampered by these pesky check containers.
